File: answers/leetcode/Kth Smallest Element in a BST.cpp
//@algorithm binary search tree
//@time complexity O(k)
//@space complexity O(height)
//@result 91 / 91 test cases passed. Status: Accepted Runtime: 20 ms Submitted: 0 minutes ago You are here! Your runtime beats 46.63% of cpp submissions.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int count = 0;
        return kthSmallest(root, count, k);
    }
    int kthSmallest(TreeNode *root, int &count, int k) {
        if (NULL == root) {
            return 0;
        }
        int left = kthSmallest(root->left, count, k);
        if (count >= k) {
            return left;
        }
        ++ count;
        if (count >= k) {
            return root->val;
        }
        int right = kthSmallest(root->right, count, k);
        return right;
    }
};

File: answers/leetcode/Kth Smallest Element in a BST/Kth Smallest Element in a BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        if (root == null) {
            return -1;
        }
        int count = this.count(root.left);
        if (count + 1 < k) {
            return this.kthSmallest(root.right, k - count - 1);
        } else if (count >= k) {
            return this.kthSmallest(root.left, k);
        } else {
            return root.val;
        }
    }
    
    private int count(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return this.count(root.left) + this.count(root.right) + 1;
    }
}